src/System.Private.CoreLib/shared/System/Runtime/Serialization/SerializationInfo.cs

```diff
@@ -23,6 +26,125 @@ public sealed class SerializationInfo
         private string _rootTypeName;
         private string _rootTypeAssemblyName;
         private Type _rootType;
+        private readonly static DeserializationToken s_emptyToken = new DeserializationToken(null);
+        
+        internal static AsyncLocal<bool> AsyncDeserializationInProgress { get; } = new AsyncLocal<bool>();
+
+        // Returns true if deserialization is currently in progress
+        public static bool DeserializationInProgress
+        {
+            [MethodImplAttribute(MethodImplOptions.NoInlining)] // Methods containing StackCrawlMark local var must be marked non-inlineable
+            get
+            {
+                if (AsyncDeserializationInProgress.Value)
+                {
+                    return true;
+                }
```

Comment:
I'm not understanding how this use of AsyncLocal works, but maybe I'm just missing the intent.

Let's say I have this:
1. StartDeserialization() called, returns a token
2. Spawn two async operations, A and B, each of which captures ExecutionContext and thus retains the true value for AsyncDeserializationInProgress.
3. Operation A Dispose's the token.
4. Operation B will forever see DeserializationInProgress as true, because even though A set AsyncDeserializationInProgress.Value to false, that only affected its flow, and even though A set DeserializationInProgress to false, Operation B is short-circuiting here and will never see tracker.DeserializationInProgress.

Reply:
The idea is that both A and B were launched during deserialization. Even if a main thread finished deserialization, B is a "part" of deserialization, so its context should always see DeserializationInProgress as true.